crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -279,74 +231,97 @@ impl Plugin for VisibilityPlugin {
-/// removed. This may impact performance if meshes with many entities are frequently
-/// reassigned/removed.
 pub fn update_bounds(
     mut commands: Commands,
     meshes: Res<Assets<Mesh>>,
-    mut mesh_reassigned: Query<
-        (Entity, &Handle<Mesh>, &mut Aabb),
-        (
-            Changed<Handle<Mesh>>,
-            Without<NoFrustumCulling>,
-            Without<NoAabbUpdate>,
-        ),
-    >,
-    mut entity_mesh_map: ResMut<EntityMeshMap>,
+    mut queries: ParamSet<(
+        Query<(&Handle<Mesh>, &Aabb), Changed<Aabb>>,
+        Query<
+            (
+                Entity,
+                &Handle<Mesh>,
+                Changed<Handle<Mesh>>,
+                Option<&mut Aabb>,
+            ),
+            (Without<NoFrustumCulling>, Without<NoAabbUpdate>),
+        >,
+    )>,
+    mut mesh_aabb_map: ResMut<MeshAabbMap>,
     mut mesh_events: EventReader<AssetEvent<Mesh>>,
-    entities_lost_mesh: RemovedComponents<Handle<Mesh>>,
+    mut changed_aabbs: Local<HashSet<Handle<Mesh>>>,
+    mut inserted_aabb_components: Local<HashSet<Handle<Mesh>>>,
 ) {
-    for entity in entities_lost_mesh.iter() {
-        entity_mesh_map.deregister(entity);
+    // Identify Aabbs added to / updated on entities with Handle<Mesh> outside of this system. The
+    // goal here is to avoid recomputing Aabbs for Meshes that have already been computed
+    // elsewhere.
+    for (mesh_handle, aabb) in queries.p0().iter() {
+        // If the Aabb component was inserted using Commands on the previous run of this system,
+        // do not consider it as Changed for this run.
+        if inserted_aabb_components.contains(mesh_handle) {
+            continue;
+        }
+        // NOTE: If two entities with the same Handle<Mesh> have different Aabbs then the last one
+        // will win here
+        mesh_aabb_map.insert(mesh_handle.clone_weak(), aabb.clone());
+        changed_aabbs.insert(mesh_handle.clone_weak());
     }
-
-    for (entity, mesh_handle, mut aabb) in mesh_reassigned.iter_mut() {
-        entity_mesh_map.deregister(entity);
-        if let Some(mesh) = meshes.get(mesh_handle) {
-            if let Some(new_aabb) = mesh.compute_aabb() {
-                entity_mesh_map.register(entity, mesh_handle);
-                *aabb = new_aabb;
+    inserted_aabb_components.clear();
+
+    // Check AssetEvents for Meshes to maintain the Handle<Mesh> -> Aabb map. If a Created or
+    // Modified event is observed for a Handle<Mesh> that is in the changed_aabbs set, then skip
+    // recomputing the Aabb.
+    let mut updated_aabbs = HashSet::new();
+    for event in mesh_events.iter() {
+        match event {
+            AssetEvent::Created { handle } | AssetEvent::Modified { handle } => {
+                if changed_aabbs.contains(handle) {
+                    // If the Aabb was manually added for this Handle<Mesh>, do not recompute it
+                    changed_aabbs.remove(handle);
```

Comment:
This case does not (always) represent "manually adding". It also covers multiple created/modified events happening in the same frame for the same mesh. This would result in skipping changes.